### PR TITLE
Cleanup of runtime test build scripts

### PR DESCRIPTION
**Pull request by trylek**
The primary objective of this change is to move most test build logic from the pair of scripts <code>src/tests/build.cmd / sh</code> to the common MSBuild project <code>src/tests/build.proj</code> to reduce maintenance cost incurred by the two scripts that need to be developed in sync, slightly increase performance (especially in small incremental scenarios) by reducing the number of separate MSBuild launches and move closer to the ability to build / run runtime tests from the root build script.

As of publishing this PR I have the change working locally in CoreCLR on Windows and Linux, from now on I need to rely much more on lab testing exercising various build combinations. I believe the change is now ready for at least general conceptual feedback regarding organization of the change. Once the testing gets clean I'll remove the WIP tag and I'll ask for a thorough review of the finalized change.

Thanks

Tomas

/cc @dotnet/runtime-infrastructure

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
I haven’t had a chance to really take a close look, but I have one request from what I’ve looked at. Can we please move the native build back into the cmd/sh scripts? Debugging the native build gets more difficult when the main entry-point is through MSBuild, and we aren’t even sharing the Windows and Linux halves of the build, so it doesn’t seem worth moving into MSBuild at the moment.

**Comment by trylek** [comment]
Thanks @jkoritzinsky for your initial feedback, I think that makes perfect sense at this point, I'll modify the change as you suggest.

**Comment by ViktorHofer** [comment]
> Can we please move the native build back into the cmd/sh scripts? Debugging the native build gets more difficult when the main entry-point is through MSBuild

IIRC @jkotas mentioned in the past that exposing only the msbuild entry point for the runtime tests should be fine.

**Comment by jkoritzinsky** [comment]
If we were to go the route of moving the CMake steps into the MSBuild projects in the future, we should use the Arcade CMake SDK that I designed for that purpose.

**Comment by fanyang-mono** [comment]
The mono mobile targets related change looks fine to me so far. @trylek If you need help investigating the failures for mono mobile targets CI lanes, feel free to let me know.

**Comment by trylek** [comment]
Thanks @fanyang-mono, I believe I have reached the stage where I have all CoreCLR tests working and some Mono tests too, there are just two remaining failure buckets I haven't yet managed to get my head around and I would definitely appreciate your help:

1) For the two arm64 test legs there's some problem with building the native components; the <code>gen-buildsys.sh</code> script claims the targeting architecture is wrong even though it does seem to get passed in the command-line argument.

2) For the two arm legs there's a configuration inconsistency - we're pulling down a debug version of the framework (just like in pre-existing PR runs without my change) but we seem to be expecting release during the LLVM AOT Compile CoreCLR tests step. Have you got any idea where we're supposed to be overriding the LibraryConfiguration property in this case? In the scripts it defaults to release and I haven't yet found out where we've been switching it over to Debug previously.

Thanks a lot

Tomas

**Comment by trylek** [comment]
As a side note, I just found out that the test wrapper build is still hosed on Linux, on Windows it's fine now. I'm about to call it a day for now, I'll investigate this tomorrow.

**Comment by trylek** [comment]
@fanyang-mono, @naricc - I believe I have fixed the most obvious infra issues regarding Mono runs but apparently some failures still remain - could you please take a quick look when you have a chance and help me understand whether some of these are known issues and / or whether there are remaining problems in my change like not passing all the necessary properties to <code>issues.targets</code> so that it doesn't filter out some tests known not to work?

Thanks

Tomas

**Comment by naricc** [comment]
@trylek So, it looks like what is going wrong is that RuntimeVariant is not being propagated to issues.targets. There are a bunch of tests that are only excluded in certain mono variants, like interpreter. And it seems like they are incorrectly still being run with those variants.

I don't see exactly what in your change is causing that problem though. I will look more tomorrow if you don't figure it out before then.

**Comment by trylek** [comment]
With Jeremy's help I have finally managed to nail down the last problem with my change - on Linux, I was generating the XUnit test wrappers twice, first in the correct wrapper generation step and also in the next <code>CORE_ROOT</code> population step which was missing the <code>RuntimeVariant</code> property. I have removed the **WIP** specifier as I believe the change is now ready to be reviewed and merged in.

Thanks

Tomas

**Comment by trylek** [comment]
Apologies about the additional churn, triggering the outerloop and R2R pipelines has opened a Pandora's box I haven't yet been able to fully address; I believe that the test priority distinction is now handled correctly but still a problem remains, some tests cannot find their build artifacts, I'm still investigating what is causing this.

**Comment by trylek** [comment]
OK, I believe I'm out of the woods now, PR is clean and so is most of outerloop and R2R run, please review when you have a chance!

File: eng/common/msbuild.sh
Comment (jkoritzinsky) [inline eng/common/msbuild.sh:57]: This change will be lost when we pull another Arcade update.

Is this a local-debugging-only change?
```suggestion
```

File: src/tests/Directory.Build.props
Comment (jkoritzinsky) [inline src/tests/Directory.Build.props:95]: This comment looks out of place. Is it still relevant?

File: src/tests/build.cmd
Comment (jkoritzinsky) [inline src/tests/build.cmd:109]: Can we add docs for this option?

File: src/tests/build.proj
Comment (jkoritzinsky) [inline src/tests/build.proj:107]: `<CallTarget>` is generally not considered good practice in MSBuild. Would it be possible to chain this in a different way (either recursive `<MSBuild />` or with target dependencies)?

File: src/tests/build.sh
Comment (jkoritzinsky) [inline src/tests/build.sh:124]: Can we change this comment to have the url to the correct issue?

File: src/tests/build.sh
Comment (jkoritzinsky) [inline src/tests/build.sh:113]: I'm personally not a fan of how MSBuild implicitly imports the current environment as properties and I generally try to avoid it if possible. Additionally, if we ever are able to move to unifying the runtime test build script entry-point into the root build scripts, this will definitely need to be cleaned up for that. This can wait until later work though as the environment variables are easily identifiable.

File: src/tests/run.proj
Comment (jkoritzinsky) [inline src/tests/run.proj:3]: Since it looks like `run.proj` doesn't import the NoTargets SDK, it won't import the Directory.Build.props/targets files. If that's intentional, it might be worthwhile to move into a subdirectory and use the NoTargets SDK + empty Directory.Build.props/targets files to make it more easily understood at a glance. (I know I've sunk a few hours here and there trying to figure out why Directory.Build.props/targets wasn't being imported when I think it should have been or vice versa).

File: src/tests/run.proj
Comment (jkoritzinsky) [inline src/tests/run.proj:852]: Same comment re CallTarget usage.

File: eng/common/msbuild.sh
Comment (trylek) [reply eng/common/msbuild.sh:57]: Oh, sorry about that one, that was indeed a temporary instrumentation helping me nail down some of the build issues, I'll revert this part of the change.

File: src/tests/Directory.Build.props
Comment (trylek) [reply src/tests/Directory.Build.props:95]: Fair enough. The comment is still relevant but it's kind of orphaned after the code shuffles, I need to find a new home for it :-).

File: src/tests/build.cmd
Comment (trylek) [reply src/tests/build.cmd:109]: Good point, added.

File: src/tests/build.sh
Comment (trylek) [reply src/tests/build.sh:124]: Thanks for spotting, fixed in updated iteration.

File: src/tests/build.sh
Comment (jkoritzinsky) [inline src/tests/build.sh:124]: As this issue has been closed, it's worth taking a look and seeing if we can remove the warnAsError flag.

Comment (trylek) [reply src/tests/build.sh]: Removed in the latest commit, let's see what the lab says.

File: src/tests/build.proj
Comment (trylek) [reply src/tests/build.proj]: Fixed in latest commit.

File: src/tests/build.sh
Comment (trylek) [reply src/tests/build.sh:113]: Yes, if you don't mind, I'd prefer to leave that to a separate follow-up change. The main problem is that escaping will be necessary at least for some of the properties (in particular I hit this with the <code>__MsgPrefix</code> and <code>__ErrMsgPrefix</code> variables) and that will require a bit more experimentation; moreover the escaping will be probably done in slightly different ways in the cmd/sh version of the script.

File: src/tests/run.proj
Comment (trylek) [reply src/tests/run.proj]: Fixed in latest commit.

File: src/tests/run.proj
Comment (trylek) [reply src/tests/run.proj:3]: I've been actually thinking about merging the run.proj script into the build.proj script; with this change it's no longer called directly, build.proj is always the "front-end"; after all, even with my additions build.proj has less than 200 lines and in combination with run.proj it will have just about 1000 lines. But I would also prefer to do that separately to make git history cleaner, merging the scripts will be a purely mechanical change.

Comment (jkoritzinsky) [reply src/tests/run.proj:3]: Sure we can hold off on that until later.

**Review by jkoritzinsky — APPROVED** [review]
LGTM barring CI failures. Looks like there might be some cases that still need `-warnAsError false` in the Pri1 test build, so we can undo that change if needed to get CI more green.

File: src/tests/build.sh
Comment (trylek) [reply src/tests/build.sh:124]: Reverted per your PR comment, I'm rerunning the outerloop job.

**Pull request merged**
